vpr/src/device/rr_spatial_lookup.cpp

```diff
@@ -0,0 +1,108 @@
+/***************************************
+ * Methods for Object RRSpatialLookup
+ ***************************************/
+#include "vtr_assert.h"
+#include "rr_spatial_lookup.h"
+
+/****************************
+ * Constructors
+ ****************************/
+RRSpatialLookup::RRSpatialLookup(t_rr_node_indices& rr_node_indices)
+    : rr_node_indices_(rr_node_indices) {
+}
+
+/****************************
+ * Accessors
+ ****************************/
+RRNodeId RRSpatialLookup::find_node(int x,
+                                    int y,
+                                    t_rr_type type,
+                                    int ptc,
+                                    e_side side) const {
+    /* Find actual side to be used */
+    e_side node_side = side;
+    if (type == IPIN || type == OPIN) {
+        VTR_ASSERT_MSG(side != NUM_SIDES, "IPIN/OPIN must specify desired side (can not be default NUM_SIDES)");
+    } else {
+        VTR_ASSERT(type != IPIN && type != OPIN);
+        node_side = SIDES[0];
+    }
+
+    /* Currently need to swap x and y for CHANX because of chan, seg convention */
+    size_t node_x = x;
+    size_t node_y = y;
+    if (CHANX == type) {
+        std::swap(node_x, node_y);
+    }
+
+    VTR_ASSERT_SAFE(3 == rr_node_indices_[type].ndims());
+
+    /* Sanity check to ensure the x, y, side and ptc are in range 
+     * Data type of rr_node_indice:
+     *   typedef std::array<vtr::NdMatrix<std::vector<int>, 3>, NUM_RR_TYPES> t_rr_node_indices;
+     * Note that x, y and side are the 1st, 2nd and 3rd dimensions of the vtr::NdMatrix
+     * ptc is in the std::vector<int>
+     */
+    if (size_t(type) >= rr_node_indices_.size()) {
+        /* Node type is out of range, return an invalid index */
```

Comment:
I'd avoid comments that just restate exactly what the code does. They will train readers to ignore comments, which is dangerous. And they use up vertical space, at least 6 lines alone here when counting also the `/* Reaching here, must...`

To further reduce vertical space usage, you can consider the this if permitted by this project's readability rules: simple return based on on a condition is more compact (and as readable) with the following:
```
   if (size_t(type) >= rr_node_indices_.size()) return RRNodeId::INVALID();
```

So, where interesting stuff is happening, we should emphasize it with code and comments, but boring things are nice when compact.

Reply:
Yes. I removed all the redundant comments. 
Regarding the 1-line if-clause, I personally do not have a strong opinion on it. I am not used to read 1-line if-clauses. Unless it is necessary, I prefer if-clauses with brackets.

Reply:
SGTM as well.